File: apigames/apigames.go
package apigames

import (
	"crypto/md5"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"log"
	"math/rand"
	"net/http"
	"net/url"
	"os"
	"strings"
	"time"
)

const BASE_URL = "https://v1.apigames.id/v2"

func GetMerchantID() string {
	return os.Getenv("APIGAMES_MERCHANT_ID")
}

func GetSecretKey() string {
	return os.Getenv("APIGAMES_SECRET_KEY")
}

func getMD5Hash(text string) string {
	hasher := md5.New()
	hasher.Write([]byte(text))
	return hex.EncodeToString(hasher.Sum(nil))
}

func GenerateSignature(refID string) string {
	return getMD5Hash(fmt.Sprintf("%s:%s:%s", GetMerchantID(), GetSecretKey(), refID))
}

func GenerateRefID() string {
	r := rand.New(rand.NewSource(time.Now().UnixNano()))
	return fmt.Sprintf("REF%d%d", time.Now().UnixNano()/int64(time.Millisecond), r.Intn(1000))
}

type ApiResponse struct {
	Success bool           `json:"success"`
	Message string         `json:"message,omitempty"`
	Data    any            `json:"data,omitempty"`
	Status  int            `json:"status,omitempty"`
}

// Cek username / nickname akun game via Apigames (ML, FF, dll.)
func CheckUsername(game, userID, serverID string) (any, error) {
	merchantID := GetMerchantID()
	secretKey := GetSecretKey()
	signature := getMD5Hash(fmt.Sprintf("%s%s", merchantID, secretKey))

	var targetGame string
	var queryUserID string

	g := strings.ToLower(strings.TrimSpace(game))
	if strings.Contains(g, "legend") || g == "ml" || g == "mlbb" {
		targetGame = "mobilelegend"
		queryUserID = fmt.Sprintf("%s%s", userID, serverID)
	} else if strings.Contains(g, "freefire") || g == "ff" {
		targetGame = "freefire"
		queryUserID = userID
	} else if strings.Contains(g, "higgs") {
		targetGame = "higgs"
		queryUserID = userID
	} else if strings.Contains(g, "cod") || g == "codm" {
		targetGame = "codm"
		queryUserID = userID
	} else if strings.Contains(g, "starrail") || g == "hsr" || strings.Contains(g, "honkai") {
		targetGame = "starrail"
		queryUserID = fmt.Sprintf("%s%s", userID, serverID)
	} else {
		// Game lain tidak didukung cek username resmi di Apigames, kembalikan bypass virtual
		return map[string]any{
			"status": 1,
			"rc":     200,
			"data": map[string]any{
				"username": fmt.Sprintf("ID %s", userID),
			},
		}, nil
	}

	apiURL := fmt.Sprintf("https://v1.apigames.id/merchant/%s/cek-username/%s", merchantID, targetGame)
	
	u, err := url.Parse(apiURL)
	if err != nil {
		return nil, err
	}

	q := u.Query()
	q.Set("user_id", queryUserID)
	q.Set("signature", signature)
	u.RawQuery = q.Encode()

	log.Printf("[APIGAMES CEK-USERNAME] Calling URL: %s\n", u.String())

	client := &http.Client{Timeout: 12 * time.Second}
	resp, err := client.Get(u.String())
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var result any
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}

	return result, nil
}

// Ambil daftar produk / katalog dari Apigames (Bypass / Optional)
func GetProducts(category string) (any, error) {
	refID := GenerateRefID()
	signature := GenerateSignature(refID)

	apiURL := fmt.Sprintf("%s/produk", BASE_URL)
	u, err := url.Parse(apiURL)
	if err != nil {
		return nil, err
	}

	q := u.Query()
	q.Set("merchant_id", GetMerchantID())
	q.Set("ref_id", refID)
	q.Set("signature", signature)
	q.Set("kategori", category)
	u.RawQuery = q.Encode()

	client := &http.Client{Timeout: 15 * time.Second}
	resp, err := client.Get(u.String())
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var result any
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}

	return result, nil
}

// Kirim transaksi top-up ke Apigames
func SendTransaction(orderNo, productID, userID, serverID string) (any, error) {
	signature := GenerateSignature(orderNo) // Gunakan orderNo sebagai refID

	apiURL := fmt.Sprintf("%s/transaksi", BASE_URL)
	u, err := url.Parse(apiURL)
	if err != nil {
		return nil, err
	}

	q := u.Query()
	q.Set("merchant_id", GetMerchantID())
	q.Set("ref_id", orderNo)
	q.Set("produk", productID)
	q.Set("tujuan", userID)
	if serverID != "" {
		q.Set("server_id", serverID)
	}
	q.Set("signature", signature)
	u.RawQuery = q.Encode()

	log.Printf("[APIGAMES TRANSAKSI] Calling URL: %s\n", u.String())

	client := &http.Client{Timeout: 30 * time.Second}
	resp, err := client.Get(u.String())
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var result any
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}

	return result, nil
}

// Cek status transaksi yang sudah dikirim
func CheckTransactionStatus(refID string) (any, error) {
	signature := GenerateSignature(refID)

	apiURL := "https://v1.apigames.id/v2/transaksi/status"
	u, err := url.Parse(apiURL)
	if err != nil {
		return nil, err
	}

	q := u.Query()
	q.Set("merchant_id", GetMerchantID())
	q.Set("ref_id", refID)
	q.Set("signature", signature)
	u.RawQuery = q.Encode()

	client := &http.Client{Timeout: 15 * time.Second}
	resp, err := client.Get(u.String())
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var result any
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}

	return result, nil
}

// Cek saldo / deposit di Apigames
func CheckBalance() (any, error) {
	merchantID := GetMerchantID()
	secretKey := GetSecretKey()
	signature := getMD5Hash(fmt.Sprintf("%s:%s", merchantID, secretKey))

	apiURL := fmt.Sprintf("https://v1.apigames.id/merchant/%s", merchantID)
	u, err := url.Parse(apiURL)
	if err != nil {
		return nil, err
	}

	q := u.Query()
	q.Set("signature", signature)
	u.RawQuery = q.Encode()

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Get(u.String())
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var result any
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}

	return result, nil
}

File: backup-node/apigames.js
// apigames.js — Integrasi Apigames.id API
const axios = require('axios');
const md5   = require('md5');

const BASE_URL     = 'https://v1.apigames.id/v2';
const MERCHANT_ID  = process.env.APIGAMES_MERCHANT_ID;
const SECRET_KEY   = process.env.APIGAMES_SECRET_KEY;

/**
 * Generate signature Apigames
 * Formula: md5(merchant_id + ":" + secret_key + ":" + ref_id)
 */
function generateSignature(ref_id) {
  return md5(`${MERCHANT_ID}:${SECRET_KEY}:${ref_id}`);
}

/**
 * Generate unique ref_id berdasarkan timestamp + random
 */
function generateRefId() {
  return `REF${Date.now()}${Math.floor(Math.random() * 1000)}`;
}

/**
 * Cek username / nickname akun game ML via Apigames
 * Endpoint: GET https://v1.apigames.id/merchant/{merchant_id}/cek-username/mobilelegend
 * Signature: md5(merchant_id + secret_key)   ← BEDA dengan transaksi!
 */
async function checkUsername({ user_id, server_id }) {
  const signature = md5(`${MERCHANT_ID}${SECRET_KEY}`);

  // URL cek-username TIDAK pakai /v2/ — beda dari endpoint transaksi
  const url = `https://v1.apigames.id/merchant/${MERCHANT_ID}/cek-username/mobilelegend`;

  const params = {
    user_id:   `${user_id}${server_id}`, // Gabungkan user ID + server ID (contoh: 532082832011)
    signature,
  };

  console.log(`\n[APIGAMES CEK-USERNAME]`);
  console.log(`  URL       : ${url}`);
  console.log(`  user_id   : ${user_id}${server_id}`);
  console.log(`  signature : ${signature}`);

  try {
    const response = await axios.get(url, { params, timeout: 12000 });
    console.log(`  RESPONSE  :`, JSON.stringify(response.data));
    return { success: true, data: response.data };
  } catch (err) {
    const errData = err.response?.data;
    const status  = err.response?.status;
    console.error(`  ERROR ${status} :`, errData || err.message);
    return {
      success: false,
      message: errData?.message || errData?.msg || err.message,
      raw: errData,
      status,
    };
  }
}

/**
 * Ambil daftar produk / katalog dari Apigames
 * Bisa difilter per kategori (misal: AGML untuk Mobile Legends)
 */
async function getProducts(category = 'AGML') {
  const ref_id    = generateRefId();
  const signature = generateSignature(ref_id);

  try {
    const response = await axios.get(`${BASE_URL}/produk`, {
      params: {
        merchant_id: MERCHANT_ID,
        ref_id,
        signature,
        kategori: category,
      },
      timeout: 15000,
    });

    return { success: true, data: response.data };
  } catch (err) {
    const msg = err.response?.data?.message || err.message;
    return { success: false, message: msg };
  }
}

/**
 * Kirim transaksi top-up ke Apigames
 */
async function sendTransaction({ order_no, product_id, user_id, server_id }) {
  const ref_id    = order_no; // gunakan order_no sebagai ref_id agar unik & mudah tracking
  const signature = generateSignature(ref_id);

  try {
    const params = {
      merchant_id: MERCHANT_ID,
      ref_id,
      produk: product_id,
      tujuan: user_id,
      server_id: server_id || '',
      signature,
    };

    const response = await axios.get(`${BASE_URL}/transaksi`, {
      params,
      timeout: 30000,
    });

    return { success: true, data: response.data };
  } catch (err) {
    const errData = err.response?.data;
    return {
      success: false,
      message: errData?.message || err.message,
      data: errData,
    };
  }
}

/**
 * Cek status transaksi yang sudah dikirim
 */
async function checkTransactionStatus(ref_id) {
  const signature = generateSignature(ref_id);

  try {
    const url = 'https://v1.apigames.id/v2/transaksi/status';
    const response = await axios.get(url, {
      params: {
        merchant_id: MERCHANT_ID,
        ref_id,
        signature,
      },
      timeout: 15000,
    });

    return { success: true, data: response.data };
  } catch (err) {
    const msg = err.response?.data?.message || err.message;
    return { success: false, message: msg };
  }
}

/**
 * Cek saldo / deposit di Apigames
 */
async function checkBalance() {
  const signature = md5(`${MERCHANT_ID}:${SECRET_KEY}`);

  try {
    const url = `https://v1.apigames.id/merchant/${MERCHANT_ID}`;
    const response = await axios.get(url, {
      params: {
        signature,
      },
      timeout: 10000,
    });

    return { success: true, data: response.data };
  } catch (err) {
    const msg = err.response?.data?.message || err.message;
    return { success: false, message: msg };
  }
}

module.exports = {
  generateSignature,
  generateRefId,
  checkUsername,
  getProducts,
  sendTransaction,
  checkTransactionStatus,
  checkBalance,
};
